servicetalk-concurrent-api/docs/modules/ROOT/pages/blocking-implementation.adoc

```diff
@@ -20,77 +21,186 @@ Everything inside ServiceTalk is somehow connected to one of the three asynchron
+unless configured to not offload ServiceTalk will offload from the Netty Event Loop `IoExecutor` as necessary in order
+to allow user code to block and will use a global executor, or user code can offload to a specific `Executor` using the
+`publishOn` operator. Consumers of the source who are certain that they will never block and can execute quickly on the
+`IoExecutor` Event Loop thread may also be able to opt-out from the ServiceTalk offloading.
+
+image::offloading.svg[Offloading]
+
+During `subscribe()` the execution will offload at the `subscribeOn()` operator and transition execution from the
+subscribing thread to an `Exeuctor` thread. The subscribing thread will be able to continue while the subscribe
+operation asynchronously continues on an `Executor` thread.
+
+The diagram shows a typical case, when a result is available at the source it will begin publication on the receiving
+event loop thread, which is owned by a Netty `IOExecutor`. Assuming that the default ServiceTalk offloading has been
+disabled, it will only offload at the `publishOn()` operator and transition execution from the event loop thread to an
+`Executor` thread. Once the publish signal is offloaded the event loop thread will be available again for executing
+other I/O tasks while the response is asynchronously processed on the `Executor` thread.
+
+=== `publishOn()` Example
+
+Using the `publishOn(executor)` operator allows processing of signals related to the source content on a different
+thread than is generating (or receiving) the content. It is the most common form of offloading used as it relates to
+the data path, specifically the `Subscription` methods.
+
+[source, java]
+----
+Collection<Integer> result = Publisher.range(1, 10) <2> <5>
+        .map(element -> element)  // non-offloaded NO-OP
+        .publishOn(publishExecutor)  <4>
+        .map(element -> element)  // offloaded NO-OP
+        .toFuture()  <3>
+        .get();  <1> <6>
+----
 
-ServiceTalk concurrency APIs defines which thread will be used by any asynchronous source for Part 2.
+<1> `toFuture().get` begins by calling `subscribe(Subscriber)`. Executing on the application thread, execution flows up
+the operator chain towards the source; `map` -> `publishOn` -> `map` -> `Publisher.range(1, 10)`.
 
-== Executor and asynchronous sources
+<2> Still executing on the application thread, `Range` will call `Subscriber.onSubscribe(Subscription)` on the
+`Subscriber`. This flows back down the operator chain, `Range` -> `map` -> `publishOn` -> `map` -> `toFuture`.
 
-ServiceTalk enforces that each asynchronous source **MUST** be associated with an `Executor` which is used to interact
-with its `Subscriber`. The below diagram illustrates the interaction between an asynchronous source, its `Subscriber`,
-its operators, and the `Executor`.
+<3> When `subscribe()` is complete, `toFuture()` will call `Subscription.request(Long.MAX_VALUE)`. This will cause
```

Comment:
`When subscribe() is complete,` -> I'm not sure what this means, considering removing this? I think the important part is `toFuture()`'s `onSubscribe(Subscription)` will call `Subscription.request(Long.MAX_VALUE)`. To be consistent with the other descriptions we should also add "execution flows up the operator chain towards the source; `map` -> `publishOn` -> `map` -> `Publisher.range(1, 10)`."